### ISSUE TITLE
policy fails with jax 0.4.36 from `IndexError: list index out of range` in backward fn

### ISSUE THREAD

**Issue by evanatyourservice** [google-deepmind/mctx#100]
Hello all!

There's a new error in mctx since JAX version 0.4.36. When running the policy improvement demo, it fails with error `IndexError: list index out of range` stemming from the backward fn. This is with python 3.11, and it runs fine with JAX version 0.4.35. I didn't see any indexing related changes in jax's change log, but maybe I'm missing something. Here is the error:

```
/opt/anaconda3/envs/jax/bin/python /Users/evanwalters/mctx/examples/policy_improvement_demo.py 
INFO:2024-12-06 01:00:48,626:jax._src.xla_bridge:927: Unable to initialize backend 'rocm': module 'jaxlib.xla_extension' has no attribute 'GpuAllocatorConfig'
I1206 01:00:48.626108 8651321920 xla_bridge.py:927] Unable to initialize backend 'rocm': module 'jaxlib.xla_extension' has no attribute 'GpuAllocatorConfig'
INFO:2024-12-06 01:00:48,627:jax._src.xla_bridge:927: Unable to initialize backend 'tpu': INTERNAL: Failed to open libtpu.so: dlopen(libtpu.so, 0x0001): tried: 'libtpu.so' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibtpu.so' (no such file), '/opt/anaconda3/envs/jax/bin/../lib/libtpu.so' (no such file), '/usr/lib/libtpu.so' (no such file, not in dyld cache), 'libtpu.so' (no such file), '/usr/local/lib/libtpu.so' (no such file), '/usr/lib/libtpu.so' (no such file, not in dyld cache)
I1206 01:00:48.627507 8651321920 xla_bridge.py:927] Unable to initialize backend 'tpu': INTERNAL: Failed to open libtpu.so: dlopen(libtpu.so, 0x0001): tried: 'libtpu.so' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibtpu.so' (no such file), '/opt/anaconda3/envs/jax/bin/../lib/libtpu.so' (no such file), '/usr/lib/libtpu.so' (no such file, not in dyld cache), 'libtpu.so' (no such file), '/usr/local/lib/libtpu.so' (no such file), '/usr/lib/libtpu.so' (no such file, not in dyld cache)
Traceback (most recent call last):
  File "/Users/evanwalters/mctx/examples/policy_improvement_demo.py", line 149, in <module>
    app.run(main)
  File "/opt/anaconda3/envs/jax/lib/python3.11/site-packages/absl/app.py", line 308, in run
    _run_main(main, args)
  File "/opt/anaconda3/envs/jax/lib/python3.11/site-packages/absl/app.py", line 254, in _run_main
    sys.exit(main(argv))
             ^^^^^^^^^^
  File "/Users/evanwalters/mctx/examples/policy_improvement_demo.py", line 135, in main
    rng_key, output = jitted_run_demo(rng_key)
                      ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/evanwalters/mctx/examples/policy_improvement_demo.py", line 73, in _run_demo
    policy_output = mctx.gumbel_muzero_policy(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/evanwalters/mctx/mctx/_src/policies.py", line 189, in gumbel_muzero_policy
    search_tree = search.search(
                  ^^^^^^^^^^^^^^
  File "/Users/evanwalters/mctx/mctx/_src/search.py", line 111, in search
    _, tree = loop_fn(
              ^^^^^^^^
  File "/Users/evanwalters/mctx/mctx/_src/search.py", line 103, in body_fun
    tree = backward(tree, next_node_index)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/evanwalters/mctx/mctx/_src/search.py", line 290, in backward
    tree, _, _ = jax.lax.while_loop(cond_fun, body_fun, loop_state)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
IndexError: list index out of range
--------------------
For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.
```

I'll try to take a closer look tomorrow, but maybe someone with more experience with mctx could help me track this error down.

**Comment by fidlej** [comment]
I was able to reproduce the issue with jax 0.4.36.

When using JAX_TRACEBACK_FILTERING=off, I can see the cause of the error.
The error is from this line:
https://github.com/jax-ml/jax/blob/2ff90382d2b046dd4474a778aced59461c30c6f7/jax/_src/lax/control_flow/loops.py#L386

Have you found more details?

**Comment by fidlej** [comment]
The bug was fixed in jax 0.4.37:
https://github.com/jax-ml/jax/releases/tag/jax-v0.4.37

**Comment by evanatyourservice** [comment]
@fidlej thank you!!
